### PR TITLE
plugins: exec: exit fluent-bit after oneshot and propagate exit code

### PR DESCRIPTION
**Pull request by ringerc**
Enable use of `fluent-bit` as a log-processing command wrapper by enhancing the `exec` plugin with options to exit on child process exit and propagate exit code.

Adds new options to `in_exec` plugin:

    Exit_After_Oneshot [bool]
    Propagate_Exit_Code [bool]

`Exit_After_Oneshot` implies` Oneshot` and causes `fluent-bit` to terminate when the command run by the `exec` plugin terminates.

`Propagate_Exit_Code` causes `fluent-bit` to exit with the same exit code as the on-shot child process it invoked, following the shell convention for handling of signal exits.

```
$ ./bin/fluent-bit -q -o stdout -i exec -p propagate_exit_code=true -p exit_after_oneshot=true -p command='echo some command output; exit 2' 
[0] exec.0: [[1681771813.322498607, {}], {"exec"=>"some command output"}]
$ echo $?
2
```

Obviously practical uses will really be parsing and transforming the logs, shipping logs to external capture services, etc.

While exiting when the one-shot command exits would probably be a sensible default, that might surprise any existing users of the current one-shot option so it is not made the default. That way this change remains backwards compatible with existing `oneshot` uses.

----
Enter `[N/A]` in the box, if an item is not applicable to your change.

**Testing**
Before we can approve your change; please submit the following in a comment:
- [x] Example configuration file for the change
- [x] Debug log output from testing the change

- [x] Attached [Valgrind](https://valgrind.org/docs/manual/quick-start.html) output that shows no leaks or memory corruption was found

If this is a change to packaging of containers or native binaries then please confirm it works for all targets.
- [ ] Run [local packaging test](./packaging/local-build-all.sh) showing all targets (including any new ones) build.
- [ ] Set `ok-package-test` label to test for all targets (requires maintainer to do).

**Documentation**

- [x] Documentation required for this feature

Docs PR: https://github.com/fluent/fluent-bit-docs/pull/1080

**Backporting**

- [ ] Backport to latest stable release.

----

Fluent Bit is licensed under Apache 2.0, by submitting this pull request I understand that this code will be released under the terms of that license.

### CODE REVIEW

**Comment by ringerc** [comment]
Config samples; these all expect a `false.sh` to exist with this content:

```
#!/bin/bash
echo "some output here"
exit 2
```

and `chmod +x false.sh`

[fluent-bit-exec-conf.txt](https://github.com/fluent/fluent-bit/files/11246241/fluent-bit-exec-conf.txt)

```
$ ./bin/fluent-bit -c fluent-bit-exec-conf.txt            
Fluent Bit v2.1.0
* Copyright (C) 2015-2022 The Fluent Bit Authors
* Fluent Bit is a CNCF sub-project under the umbrella of Fluentd
* https://fluentbit.io

[2023/04/17 16:40:05] [ info] [fluent bit] version=2.1.0, commit=9efd3532a8, pid=2061089
[2023/04/17 16:40:05] [ info] [storage] ver=1.4.0, type=memory, sync=normal, checksum=off, max_chunks_up=128
[2023/04/17 16:40:05] [ info] [cmetrics] version=0.6.1
[2023/04/17 16:40:05] [ info] [ctraces ] version=0.3.0
[2023/04/17 16:40:05] [ info] [input:exec:exec.0] initializing
[2023/04/17 16:40:05] [ info] [input:exec:exec.0] storage_strategy='memory' (memory only)
[2023/04/17 16:40:05] [ info] [sp] stream processor started
[2023/04/17 16:40:05] [ info] [output:stdout:stdout.0] worker #0 started
[0] stdin: [[1681706405.178401054, {}], {"exec"=>"some output here"}]
^C[2023/04/17 16:38:36] [engine] caught signal (SIGINT)
```

Runs until signalled. Exiting behaviour.

[fluent-bit-exec-and-exit-conf.txt](https://github.com/fluent/fluent-bit/files/11246236/fluent-bit-exec-and-exit-conf.txt)

```
$ ./bin/fluent-bit -c fluent-bit-exec-and-exit-conf.txt 
Fluent Bit v2.1.0
* Copyright (C) 2015-2022 The Fluent Bit Authors
* Fluent Bit is a CNCF sub-project under the umbrella of Fluentd
* https://fluentbit.io

[2023/04/17 16:40:17] [ info] [fluent bit] version=2.1.0, commit=9efd3532a8, pid=2061175
[2023/04/17 16:40:17] [ info] [storage] ver=1.4.0, type=memory, sync=normal, checksum=off, max_chunks_up=128
[2023/04/17 16:40:17] [ info] [cmetrics] version=0.6.1
[2023/04/17 16:40:17] [ info] [ctraces ] version=0.3.0
[2023/04/17 16:40:17] [ info] [input:exec:exec.0] initializing
[2023/04/17 16:40:17] [ info] [input:exec:exec.0] storage_strategy='memory' (memory only)
[2023/04/17 16:40:17] [ info] [sp] stream processor started
[2023/04/17 16:40:17] [ info] [output:stdout:stdout.0] worker #0 started
[2023/04/17 16:40:17] [ warn] [engine] service will shutdown in max 5 seconds
[0] stdin: [[1681706417.285810237, {}], {"exec"=>"some output here"}]
[2023/04/17 16:40:18] [ info] [engine] service has stopped (0 pending tasks)
[2023/04/17 16:40:18] [ info] [output:stdout:stdout.0] thread worker #0 stopping...
[2023/04/17 16:40:18] [ info] [output:stdout:stdout.0] thread worker #0 stopped
$ echo $?                                              
0
```

Exits once command exits. Even though command exited with nonzero, fluent-bit exits with zero.

[fluent-bit-exec-and-exit-propagate-conf.txt](https://github.com/fluent/fluent-bit/files/11246259/fluent-bit-exec-and-exit-propagate-conf.txt)

```
./bin/fluent-bit -c fluent-bit-exec-and-exit-propagate-conf.txt
Fluent Bit v2.1.0
* Copyright (C) 2015-2022 The Fluent Bit Authors
* Fluent Bit is a CNCF sub-project under the umbrella of Fluentd
* https://fluentbit.io

[2023/04/17 16:43:53] [ info] [fluent bit] version=2.1.0, commit=9efd3532a8, pid=2062027
[2023/04/17 16:43:53] [ info] [storage] ver=1.4.0, type=memory, sync=normal, checksum=off, max_chunks_up=128
[2023/04/17 16:43:53] [ info] [cmetrics] version=0.6.1
[2023/04/17 16:43:53] [ info] [ctraces ] version=0.3.0
[2023/04/17 16:43:53] [ info] [input:exec:exec.0] initializing
[2023/04/17 16:43:53] [ info] [input:exec:exec.0] storage_strategy='memory' (memory only)
[2023/04/17 16:43:53] [ info] [sp] stream processor started
[2023/04/17 16:43:53] [ info] [output:stdout:stdout.0] worker #0 started
sh: 1: false.sh: not found
[2023/04/17 16:43:53] [ warn] [engine] service will shutdown in max 5 seconds
[2023/04/17 16:43:54] [ info] [engine] service has stopped (0 pending tasks)
[2023/04/17 16:43:54] [ info] [output:stdout:stdout.0] thread worker #0 stopping...
[2023/04/17 16:43:54] [ info] [output:stdout:stdout.0] thread worker #7085 
$ echo $?
127
```

Exits when command exits, propagates child process exit code.

**Comment by ringerc** [comment]
Open question: is handling of child proc exiting on signal correct?

**Comment by ringerc** [comment]
Updated PR to improve signal handling logic and related comments.

I can't see any existing tests for the `exec` plugin. I'd be happy to extend tests to cover the new features but I don't know how to write a fluent-bit plugin test from scratch.

**Comment by ringerc** [comment]
I went to run this with `valgrind` but I see a *lot* of `valgrind` errors on an *unpatched* `fluent-bit` using the `valgrind.supp` shipped with `fluent-bit`, so I'm not sure how useful it'll be. It is a very unclean run with a lot of noise within `fluent-bit` itself - sufficiently so that I couldn't write a sensible suppressions file in a reasonable time that wouldn't land up hiding anything within the plugin invocation as well.

Built on Ubuntu 22.10 with gcc 12.2.0 using default cmake options:

```
➜  fluent-bit git:(master) ✗ valgrind -s --gen-suppressions=all --leak-check=full --suppressions=./valgrind.supp --suppressions=/usr/lib/valgrind/debian.supp --suppressions=/usr/libexec/valgrind/default.supp --max-stackframe=28681792 ./bin/fluent-bit -q -o stdout -i exec -p oneshot=true -p command='echo some command output; exit 2' >& valgrind.out

➜  fluent-bit git:(master) ✗ wc -l valgrind.out 
11186 valgrind.out
➜  fluent-bit git:(master) ✗ tail -1 valgrind.out 
==2117629== ERROR SUMMARY: 14623 errors from 543 contexts (suppressed: 0 from 0)
```

I used the Perl script in https://wiki.wxwidgets.org/Parse_valgrind_suppressions.sh to generate a suppressions file; obviously hand-writing would be better, but it was just too huge. With a number of iterations I ended up with a suppressions file that matched for the simple case I ran, but since some of the suppressions are calls with variable paths into libraries, I found that the tiniest change in invocation led to new and unrelated valgrind errors.

Unless there's a current suppressions file somewhere, or some build/runtime options that encourage fluent-bit to run more valgrind-clean, I can't give you any useful valgrind output. All I can really do is attach the suppressions file that matches all the valgrind errors I see reported, which I will include here.

[valgrind_supp.txt](https://github.com/fluent/fluent-bit/files/11257115/valgrind_supp.txt)

Of these suppressions, the only ones that mention the `in_exec` plugin are calls in `__pthread_disable_asynccancel` or `__pthread_enable_asynccancel`, which I see widespread throughout runs on `main` too. They don't look relevant, and appear intermittently.

I'll try to attach a full valgrind run log anyway, but the noise is so great I don't expect it'll be much use.


[valgrind_out.txt](https://github.com/fluent/fluent-bit/files/11257133/valgrind_out.txt)

**Comment by ringerc** [comment]
[fluent-bit-exec-and-exit-propagate-debuglog.txt](https://github.com/fluent/fluent-bit/files/11257141/fluent-bit-exec-and-exit-propagate-debuglog.txt)

Debug logs with the propagate flag set, run as

```
$ ./bin/fluent-bit -c fluent-bit-exec-and-exit-propagate-conf.txt -vvv >& fluent-bit-exec-and-exit-propagate-debuglog.txt
```

**Comment by ringerc** [comment]
What're the next steps here?

I'm currently working around this with a really ugly hack using the `stdin` plugin, `fluent-bit` running as a co-process with file descriptor redirection, etc.

**Comment by ringerc** [comment]
Is this project active? Is there anything else I need to do in order to get a PR checked?

**Review by leonardo-albertovich — APPROVED** [review]

**Comment by leonardo-albertovich** [comment]
Enabling the exec input plugin is something planned on doing soon but I noticed the macros to check the process status might not be available yet something easy to overcome. Would you mind taking a look at that? I'd really appreciate it.

**Comment by leonardo-albertovich** [comment]
Btw, sorry about the lack of interaction, for some reason I didn't see this PR before.

**Comment by ringerc** [comment]
@leonardo-albertovich 

> but I noticed the macros to check the process status might not be available yet

I'm not sure what you mean by this. Do you mean the macros `WIFEXITED`, `WEXITSTATUS`, `WIFSIGNALED` and `WTERMSIG` ? They're provided by `sys/wait.h` and are specified by POSIX.1-2001 and have been around absolutely forever. See https://support.sas.com/documentation/onlinedoc/ccompiler/doc750/html/lr2/z2005123.htm .

Are you asking for a CMake feature test for these macros to be added? I could probably do that if required, though it's kind of implied by building on a posix-alike platform.

Is your concern re MS Windows support? I didn't find a lot of info about Windows in the fluent-bit docs, but `cmake/windows-setup.cmake` lists the `FLB_IN_EXEC` plugin as not supported on Windows anyway. It would require major changes to work on Windows, the least of which would be process exit status macros.

---

I took a look at the failing test runs, but there's no sign they're relevant to this change as opposed to being unstable tests:

* https://github.com/fluent/fluent-bit/actions/runs/4726708830/jobs/8474537865?pr=7207 failed for `flb-rt-in_event_test` and `flb-rt-out_tcp` which are not touched in any way in this PR.
* https://github.com/fluent/fluent-bit/actions/runs/4726708830/jobs/8474537996?pr=7207 failed for `flb-rt-core-timeout`, `flb-rt-in_event_test` and `flb-rt-out_tcp` and I don't see how those could be related to this PR's changes
* https://github.com/fluent/fluent-bit/actions/runs/4726708830/jobs/8474538119?pr=7207 failed for `flb-rt-core_chunk_trace`, `flb-rt-in_random` and `flb-rt-out_tcp`. Same deal.

All failures are on OSX. Looks like issues with OSX tests rather than this PR.

So I'm not quite sure what you're asking me to do.

**Comment by leonardo-albertovich** [comment]
What I was saying was that those macros are not present in windows and since the plugin already uses popen we were thinking of enabling it in windows even if that meant making it experimental and adding a big red label.

I understand the limitations and caveats that come with windows and a "universal" approach and at some point we might want to abstract the process management and console redirection in platform specific implementations shared by the same plugin but that's not what we had in mind for the moment.

So what I'm asking you to do is to make sure this feature doesn't make it harder for us to enable the plugin in windows and the two options are : 

1. Read about those macros and how they are handled in windows by others (I found a thread in an old postgresql mailing list about it) and proceed accordingly (probably defining them locally for windows)
2. Wrap lines 183-213 in an preprocessor conditional and just set `flb_exit_code` to zero in windows which considering the fact that the plugin used to completely ignore the result wouldn't be a huge deviation (from a preexisting bad behavior I guess)

**Comment by dkmorgan** [comment]
> I'm currently working around this with a really ugly hack using the `stdin` plugin, `fluent-bit` running as a co-process with file descriptor redirection, etc.

Hi @ringerc,
I would like to use your workaround until the PR gets merged.
Can you possibly share how you actually implemented your hack?

**Comment by ringerc** [comment]
@leonardo-albertovich  OK, that makes sense, so it _is_ about windows. I'll take a look.

**Comment by ringerc** [comment]
@dkmorgan something like

```
#!/usr/bin/env python3

import sys
import subprocess
import logging
import shlex

logger = logging.getLogger('main')

def main():
    "run child proc and redirect output via fluent-bit"

    child_cmd = sys.argv[1:]

    fluent_bit = subprocess.Popen(
            ["/opt/fluent-bit/bin/fluent-bit", "--config", "fluent-bit.conf", "--parser", "parsers.conf"],
            stdin=subprocess.PIPE)

    fluent_bit.stdin.write(b"Starting process " + shlex.join(child_cmd).encode(sys.getdefaultencoding()) + b"\n")
    fluent_bit.stdin.flush()

    try:
        child_proc = subprocess.Popen(child_cmd, stdout=fluent_bit.stdin, stderr=fluent_bit.stdin)
        exitcode = child_proc.wait()
        fluent_bit.stdin.write(f"Child process exited with {exitcode} \n".encode(sys.getdefaultencoding()))
        fluent_bit.stdin.flush()
    except Exception as ex:
        fluent_bit.stdin.write(b"Failed to invoke child process: " + str(ex).encode(sys.getdefaultencoding()) + b"\n")
        fluent_bit.stdin.flush()
        exitcode=1

    try:
        fluent_bit.stdin.close()
    except Exception:
        # Swallow so we can continue to wait for fluent-bit to exit. There's a timeout, so
        # we don't rely on stdin having closed successfully to avoid hanging forever.
        logger.warn("Exception while closing fluent-bit stdin")

    logger.info('waiting for fluentbit')
    try:
        fluent_bit.wait(timeout=60)
    except subprocess.TimeoutExpired:
        logger.info("Timed out waiting for fluentbit to exit")
        try:
            fluent_bit.terminate()
        except Exception:
            logger.warn("Exception while terminating fluent-bit")

    # bubble up the exit code from the wrapped command
    logger.info('fluentbit exited: %s', fluent_bit.returncode)
    exitcode=child_proc.returncode
    if exitcode < 0:
        # translate python subprocess negative signal exits to shell-convention
        # 128+N signal exits; see https://www.gnu.org/software/bash/manual/html_node/Exit-Status.html
        exitcode = 128-child_proc.returncode
    sys.exit(exitcode)

if __name__ == '__main__':
    logging.basicConfig(level='INFO')
    main()
```

with `fluent-bit.conf` (obviously you'll use a different one, this is just the bare minimum):

```
[INPUT]
    # Override the stdin plugin's default json parsing by reading stdin line-by-line as raw strings instead of default (which is json)
    Name    stdin
    Tag     fluentbit_stdin
    Parser  stringify_message

[OUTPUT]
    Name   stdout
    Match  *
```

and `parsers.conf`

```
[PARSER]
# We initially read stdin as a single message with key "message".
# If this parser isn't configured, the stdin plugin will try to read the input as json and fail.
    name        stringify_message
    format      regex
    Key_Name    message
    regex       ^(?<message>.*)
```

I'm sure you can see why I bothered with the `exec` plugin changes.

Note that this approach does nothing special for signal handling. A `SIGTERM` is usually delivered to the whole process tree. AFAICS by default `fluent-bit` ignores the first `SIGTERM` it receives, so this tends to work ok. The script should possibly trap `SIGTERM` and ensure that it still waits for `fluent-bit` to exit, but I have not bothered fiddling with that.

**Comment by leonardo-albertovich** [comment]
Thanks @ringerc, we appreciate the effort. Please let me know once you make those changes and I'll merge this.

**Comment by ringerc** [comment]
Test steps

* Deploy Azure VM https://azuremarketplace.microsoft.com/en/marketplace/apps/microsoftvisualstudio.visualstudio2022?tab=PlansAndPrice - I used "Visual Studio 2022 Community on Windows Server 2022 (x64)"
* Install CMake
* git clone my branch
* Ignore all dependencies since I only care about the core and `in_exec` plugin for this
* `cmake .` in my clone
* wait an eternity for sloooow windows build tooling to get around to checks
* `cmake --build .`

**Comment by ringerc** [comment]
OK, having got this far, I see that the `exec` plugin isn't built, so I have to patch the cmake build to enable it. In `plugins/CMakeLists.txt` I see that `in-exec` is guarded by an `IF (FLB_PARSER)` directive. Which seems to be on by default. 

It looks like the problem is likely from `cmake/windows-setup.cmake`

```
if (NOT (${FLEX_FOUND} AND ${BISON_FOUND}))
  message(STATUS "flex and bison not found. Disable stream_processor building.")
  set(FLB_STREAM_PROCESSOR No)
  set(FLB_RECORD_ACCESSOR  No)
endif()
```

which builds a limited fluent-bit instead of failing if it doesn't find flex and bison. Which is weird, because I put flex and bison on the path.

It looks like this is because the cmake cache remembers that flex and bison are absent. But other parts of the build actually rely on flex and bison being present and will fail if they're not present (making the whole check pointless anyway). So it _appears_ like adding them will cause fluentbit to built with them, but in reality a weird build is produced that uses flex and bison but doesn't enable all features that require flex and bison.

I had to nuke the cmake cache and wait for the incredibly slow initial discovery again.

I landed up patching `cmake/windows-setup.cmake` to `MESSAGE(FATAL_ERROR ...)` inside the `if (NOT (${FLEX_FOUND} AND ${BISON_FOUND}))` clause then running

```
mkdir ..\fluent-bit-build
cd ..\fluent-bit-build
cmake -Wno-dev -S ..\fluent-bit -B .  -DOPENSSL_ROOT_DIR='C:\Program Files\OpenSSL-Win64'
cmake --build .
```

**Comment by ringerc** [comment]
Finally I get to the point where the windows built actually attempts to compile the plugin. Failing as expected with

```
fluent-bit\plugins\in_exec\in_exec.c(32,10): fatal  error C1083: Cannot open include file: 'sys/wa
it.h': No such file or directory
```

**Comment by ringerc** [comment]
OK, so I had to figure out how to build fluentbit on win32, write a patch to support win32 in `in_exec.c` since it cannot be built for win32 as it stands, then port my patch over to win32 as well. As usual, Windows makes everything awful.

I created a file `false.ps1`

```
echo "line1"
echo "line2"
exit 1
```

and a fluent-bit config `exec-exit-and-propagate-conf.txt` for win32 that's the same as above, except the command is `powershell false.ps1`:

```
[INPUT]
    Name          exec
    Tag           stdin
    Command       powershell .\false.ps1
    Propagate_Exit_Code True
    Exit_After_Oneshot true

[OUTPUT]
    Name   stdout
    Match  *
```

and confirmed it worked on win32

```
PS C:\Users\ringerc\dev\fluent-bit-build> .\bin\Debug\fluent-bit.exe -c .\exec-and-propagate-conf.txt
Fluent Bit v2.1.3
* Copyright (C) 2015-2022 The Fluent Bit Authors
* Fluent Bit is a CNCF sub-project under the umbrella of Fluentd
* https://fluentbit.io

[2023/05/22 03:09:14] [ info] [fluent bit] version=2.1.3, commit=fa9c6977a7, pid=7276
[2023/05/22 03:09:14] [ info] [storage] ver=1.4.0, type=memory, sync=normal, checksum=off, max_chunks_up=128
[2023/05/22 03:09:14] [ info] [cmetrics] version=0.6.1
[2023/05/22 03:09:14] [ info] [ctraces ] version=0.3.1
[2023/05/22 03:09:14] [ info] [input:exec:exec.0] initializing
[2023/05/22 03:09:14] [ info] [input:exec:exec.0] storage_strategy='memory' (memory only)
[2023/05/22 03:09:14] [ info] [sp] stream processor started
[2023/05/22 03:09:14] [ info] [output:stdout:stdout.0] worker #0 started
[0] stdin: [[[2023/05/22 03:09:14] [ warn] [engine] service will shutdown in max 5 seconds
1684724954.760278600, {}], {"exec"=>"line1"}]
[1] stdin: [[1684724954.760536800, {}], {"exec"=>"line2"}]
[2023/05/22 03:09:15] [ info] [engine] service has stopped (0 pending tasks)
[2023/05/22 03:09:15] [ info] [output:stdout:stdout.0] thread worker #0 stopping...
[2023/05/22 03:09:15] [ info] [output:stdout:stdout.0] thread worker #0 stopped
PS C:\Users\ringerc\dev\fluent-bit-build>
```

though the exit code isn't right on windows yet

```
PS C:\Users\ringerc\dev\fluent-bit-build> echo $LASTEXITCODE
0
```

I honestly don't care about that yet.

I've pushed the required changes to this branch, including the new cmake tests etc.

**Comment by ringerc** [comment]
OK, fixed the exit code. MSDN documentation was misleading. Shock, no, really?

**Comment by ringerc** [comment]
@leonardo-albertovich I've revised this to build on windows. It was a massive pain honestly - not porting the patch, but getting fluent-bit to build on windows so I could port and test the patch. See my notes above. Please consider changing `cmake/windows-setup.cmake` so that the block under `if (NOT (${FLEX_FOUND} AND ${BISON_FOUND}))` just aborts with something like `message(fatal_error "flex and bison required for win32 builds, see DEVELOPER_GUIDE.md")` and a section for windows in there. Even if it just points to the `appveyor.yml` or the `.github/workflows/call-build-windows.yaml` or the `dockerfiles/Dockerfile.windows` or whatever should be used as a reference to kludge together a working windows build environment.

In any case, I've pushed changes to:

* add a cmake check for `sys/wait.h`
* abstract the exit status macros, popen and pclose calls in a `plugins/in_exec/in_exec_win32_compat.h` file, hiding the win32 gore
* use the abstracted flavours in `plugins/in_exec/in_exec.c`

and I've tested that it works on win32 using the config in the above comment:

```
PS C:\Users\ringerc\dev\fluent-bit-build> .\bin\Debug\fluent-bit.exe -c .\exec-and-propagate-conf.txt -v
Fluent Bit v2.1.3
* Copyright (C) 2015-2022 The Fluent Bit Authors
* Fluent Bit is a CNCF sub-project under the umbrella of Fluentd
* https://fluentbit.io

[2023/05/22 03:28:40] [ info] Configuration:
[2023/05/22 03:28:40] [ info]  flush time     | 1.000000 seconds
[2023/05/22 03:28:40] [ info]  grace          | 5 seconds
[2023/05/22 03:28:40] [ info]  daemon         | 0
[2023/05/22 03:28:40] [ info] ___________
[2023/05/22 03:28:40] [ info]  inputs:
[2023/05/22 03:28:40] [ info]      exec
[2023/05/22 03:28:40] [ info] ___________
[2023/05/22 03:28:40] [ info]  filters:
[2023/05/22 03:28:40] [ info] ___________
[2023/05/22 03:28:40] [ info]  outputs:
[2023/05/22 03:28:40] [ info]      stdout.0
[2023/05/22 03:28:40] [ info] ___________
[2023/05/22 03:28:40] [ info]  collectors:
[2023/05/22 03:28:40] [ info] [fluent bit] version=2.1.3, commit=213c2930db, pid=3492
[2023/05/22 03:28:40] [debug] [engine] coroutine stack size: 98302 bytes (96.0K)
[2023/05/22 03:28:40] [ info] [storage] ver=1.4.0, type=memory, sync=normal, checksum=off, max_chunks_up=128
[2023/05/22 03:28:40] [ info] [cmetrics] version=0.6.1
[2023/05/22 03:28:40] [ info] [ctraces ] version=0.3.1
[2023/05/22 03:28:40] [ info] [input:exec:exec.0] initializing
[2023/05/22 03:28:40] [ info] [input:exec:exec.0] storage_strategy='memory' (memory only)
[2023/05/22 03:28:40] [debug] [exec:exec.0] created event channels: read=784 write=796
[2023/05/22 03:28:40] [debug] [input:exec:exec.0] exit_after_oneshot implies oneshot mode, enabling
[2023/05/22 03:28:40] [debug] [input:exec:exec.0] interval_sec=-1 interval_nsec=-1 oneshot=1 buf_size=4096
[2023/05/22 03:28:40] [debug] [stdout:stdout.0] created event channels: read=808 write=812
[2023/05/22 03:28:40] [ info] [sp] stream processor started
[2023/05/22 03:28:40] [ info] [output:stdout:stdout.0] worker #0 started
[2023/05/22 03:28:41] [debug] [input chunk] update output instances with new chunk size diff=41
[2023/05/22 03:28:41] [debug] [input chunk] update output instances with new chunk size diff=41
[2023/05/22 03:28:41] [debug] [input:exec:exec.0] command powershell .\false.ps1 exited with code 1
[2023/05/22 03:28:41] [debug] [input:exec:exec.0] one-shot command exited, terminating fluent-bit
[0] stdin: [[1684726121.199482200, {}], {"exec"=>"line1"}]
[2023/05/22 03:28:41] [debug] [task] created task=000002763D45FA20 id=0 OK
[1] stdin: [[1684726121.199757500, {}], {"exec"=>"line2"}]
[2023/05/22 03:28:41] [debug] [output:stdout:stdout.0] task_id=0 assigned to thread #0
[2023/05/22 03:28:41] [ warn] [engine] service will shutdown in max 5 seconds
[2023/05/22 03:28:41] [debug] [out flush] cb_destroy coro_id=0
[2023/05/22 03:28:41] [debug] [task] destroy task=000002763D45FA20 (task_id=0)
[2023/05/22 03:28:42] [ info] [engine] service has stopped (0 pending tasks)
[2023/05/22 03:28:42] [ info] [output:stdout:stdout.0] thread worker #0 stopping...
[2023/05/22 03:28:42] [ info] [output:stdout:stdout.0] thread worker #0 stopped
PS C:\Users\ringerc\dev\fluent-bit-build> echo $LASTEXITCODE
1
PS C:\Users\ringerc\dev\fluent-bit-build>
```

File: cmake/platform_feature_checks.cmake [hunk 1]
```diff
@@ -0,0 +1,8 @@
+# Feature tests for various platform and compiler capabilities,
+# system headers, etc.
+
+include(CheckIncludeFile)
+CHECK_INCLUDE_FILE("sys/wait.h" FLB_HAVE_SYS_WAIT_H)
+if (FLB_HAVE_SYS_WAIT_H)
+	FLB_DEFINITION(FLB_HAVE_SYS_WAIT_H)
```
Comment (leonardo-albertovich) [inline cmake/platform_feature_checks.cmake:7]: Please indent with spaces.

File: plugins/in_exec/in_exec.c [hunk 2]
```diff
@@ -165,7 +168,66 @@ static int in_exec_collect(struct flb_input_instance *ins,
...
+         * translation is already done for us.
+         * For references on exit code handling in wrappers see
+         * https://www.gnu.org/software/bash/manual/html_node/Exit-Status.html
+         * and
+         * https://skarnet.org/software/execline/exitcodes.html
+         */
+        cmdret = FLB_PCLOSE(cmdp);
+        if (cmdret == -1) {
+            flb_errno();
+            flb_plg_debug(ctx->ins, "unexpected error while waiting for exit of command %s ", ctx->cmd);
```
Comment (leonardo-albertovich) [inline plugins/in_exec/in_exec.c:184]: Please break this and other lines that exceed the 80 column limit.

File: plugins/in_exec/in_exec_win32_compat.h [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+#define FLB_WEXITSTATUS(status) WEXITSTATUS((status))
+#define FLB_IFSIGNALED(status) WIFSIGNALED((status))
+#define FLB_WTERMSIG(status) WTERMSIG((status))
+#else
+#define FLB_WIFEXITED(status) (1)
+#define FLB_WEXITSTATUS(status) ((status) & 0x00ff)
+#define FLB_IFSIGNALED(status) (0)
+#define FLB_WTERMSIG(status) (-1)
+
+#if !defined(FLB_SYSTEM_WINDOWS)
```
Comment (leonardo-albertovich) [inline plugins/in_exec/in_exec_win32_compat.h:57]: Please define these as static inline functions instead of macros and use lowercase names.

**Review by leonardo-albertovich — CHANGES_REQUESTED** [review]

**Comment by leonardo-albertovich** [comment]
Those are all very valid points @ringerc, and I've raised the same concern about bison and flex in the past.

I'm sorry about the frustration this caused you but I'm also thankful for the effort you put on it and for the fact that you went the extra mile to accomodate to my requests.

Once you make the minimal changes I requested this will be ready to be merged. 

Thank you very much!

File: plugins/in_exec/in_exec_win32_compat.h (hunk 3 above)
Comment (leonardo-albertovich) [reply plugins/in_exec/in_exec_win32_compat.h:57]: This goes just for `FLB_POPEN` and `FLB_PCLOSE`.

**Comment by ringerc** [comment]
@leonardo-albertovich I've addressed your requested changes.

I also improved the comments in the popen/pclose win32 compat wrapper.

And I promoted one log message from `flb_plg_debug` to `flb_plg_info`:

```
[2023/05/24 10:30:35] [ info] [input:exec:exec.0] one-shot command exited, terminating fluent-bit
```

because without it, the output on default verbosity doesn't tell the reader why the engine exited, like

```
[2023/05/24 10:30:35] [ info] [output:stdout:stdout.0] worker #0 started
[2023/05/24 10:30:35] [ warn] [engine] service will shutdown in max 5 seconds
[2023/05/24 10:30:35] [ info] [engine] service has stopped (0 pending tasks)
```

and I thought it would be confusing. I have _not_ added the exit code to the message above because it was going to be ugly to do so while handling the different exit cases; just run with `-v` if you want the raw exit status of the underlying command.

Note that the `flb_pclose()` wrapper has the return value semantics of `_pclose()`, it does not try to adapt the return value to conform with posix `pclose()`. The difference is handled in the `FLB_WEXITSTATUS` and related macros. I considered having `flb_pclose()` adapt the return value to posix-like instead, but I am not sure there's any point.

**Comment by ringerc** [comment]
Relevant parts of the diff between last and this push, since rebasing onto current `master` added lots of noise:

* https://github.com/fluent/fluent-bit/compare/213c2930db66d74c73a55ef6a0bda52ac8f598fb..672286177591246af2dbde5db234a787d1712c39#diff-241ec0a5f832eb9d14a6f243a753335683ea4bcdcb73e0ba66a0377a9a01f80e
* https://github.com/fluent/fluent-bit/compare/213c2930db66d74c73a55ef6a0bda52ac8f598fb..672286177591246af2dbde5db234a787d1712c39#diff-7dc2a1ef2ad5d65f0a6cd55f5d173874ec2cf7b6f7593da12985492e4d2798dc

Note two fixes - renamed `FLB_IFSIGNALED` to `FLB_WIFSIGNALED`, and misplaced `#endif`.

**Review by leonardo-albertovich — DISMISSED** [review]

File: plugins/in_exec/in_exec_win32_compat.h [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+ * are misleading; they say that "The format of the return value is the same as
+ * for _cwait, except the low-order and high-order bytes are swapped." But
+ * _cwait isn't documented as having any meaningful return on success, the
+ * process exit code is meant to be in  its "termstat" out parameter per
+ * https://learn.microsoft.com/en-us/cpp/c-runtime-library/reference/cwait?view=msvc-170
+ * The return code of _pclose() actually appears to be the process exit code
+ * without the bit-shift that waitpid() applies.
+ */
+static inline int flb_pclose(FILE *stream) {
+	return _pclose(stream);
```
Comment (leonardo-albertovich) [inline plugins/in_exec/in_exec_win32_compat.h:90]: Sorry, I missed this when I approved the PR, could you replace TABs with 4 spaces?

**Comment by leonardo-albertovich** [comment]
Awesome, other than replacing those tabs for spaces this PR is tip top. 

Thank you very much!

**Comment by ringerc** [comment]
New separate PR to fix the windows build complaints somewhat https://github.com/fluent/fluent-bit/pull/7463

**Comment by ringerc** [comment]
@leonardo-albertovich Fixed tabs

**Comment by ringerc** [comment]
(Don't like that this PR doesn't add test cover, but I'm not writing a whole new test module for a plugin that has none already, especially given that fluent-bit uses its own bespoke test suite with limited docs and examples.)

File: plugins/in_exec/in_exec_win32_compat.h (hunk 4 above)
Comment (ringerc) [reply plugins/in_exec/in_exec_win32_compat.h]: Fixed

**Review by patrick-stephens — APPROVED** [review]

**Comment by ringerc** [comment]
@leonardo-albertovich 

The failing tests are

```

The following tests FAILED:
	  3 - flb-rt-in_event_test (SEGFAULT)
	  9 - flb-rt-in_statsd (Failed)
	 10 - flb-rt-in_syslog (Failed)
```

in https://github.com/fluent/fluent-bit/actions/runs/5063344577/jobs/9108776973?pr=7207 which are clearly unrelated to this change, and I've seen be unstable in multiple recent runs.

AFAICS this is ready for merge, how to I get it progressed to merge?

**Comment by leonardo-albertovich** [comment]
Don't worry about those, we are in the process of fixing them (it's a localized issue in macos). Your PR is fine.

**Comment by ringerc** [comment]
Thankyou @edsiper

**Comment by ringerc** [comment]
@edsiper The docs PR for this change at https://github.com/fluent/fluent-bit-docs/pull/1080 pending merge too

**Comment by patrick-stephens** [comment]
Docs will be merged once a release is done @ringerc - otherwise it confuses people to see the docs up for something not available in the release.

**Comment by patrick-stephens** [comment]
> Thankyou @edsiper
> 
> You can probably enable the `in_exec` plugin for Windows now too, since this PR ported it over to build on Windows.

Can you submit a PR for this?

**Comment by ringerc** [comment]
@patrick-stephens https://github.com/fluent/fluent-bit/pull/7664 as requested . I've torn down my Windows dev-env (it was an Azure VM) after this work closed out, so I can't easily do any manual testing of it, but the github CI stuff should run to validate the build at least.
